topdown/graphql.go

```diff
@@ -169,46 +255,60 @@ func pruneIrrelevantGraphQLASTNodes(value ast.Value) ast.Value {
 
 // Reports errors from parsing/validation.
 func builtinGraphQLParse(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
-	// Get the raw strings from each operand so that we can
-	// feed them to the GraphQL parser functions.
-	rawQuery, err := builtins.StringOperand(operands[0].Value, 1)
-	if err != nil {
-		return err
-	}
-	rawSchema, err := builtins.StringOperand(operands[1].Value, 1)
-	if err != nil {
-		return err
+	var queryDoc *gqlast.QueryDocument
+	var schemaDoc *gqlast.SchemaDocument
+	var err error
+
+	// Parse/translate query if it's a string/object.
+	switch x := operands[0].Value.(type) {
+	case ast.String:
+		queryDoc, err = stringToQueryDocument(string(x))
+		if err != nil {
+			return err
+		}
+	case ast.Object:
+		queryDoc, err = objectToQueryDocument(x)
+		if err != nil {
+			return err
+		}
+	default:
+		// Error if wrong type.
+		return builtins.NewOperandTypeErr(0, x, "string", "object")
 	}
 
-	// Generate ASTs/errors for the GraphQL query.
-	queryDoc, err := parseQuery(string(rawQuery))
-	if err != nil {
-		return err
+	// Parse/translate schema if it's a string/object.
+	switch x := operands[1].Value.(type) {
+	case ast.String:
+		schemaDoc, err = stringToSchemaDocument(string(x))
+		if err != nil {
+			return err
+		}
+	case ast.Object:
+		schemaDoc, err = objectToSchemaDocument(x)
+		if err != nil {
+			return err
+		}
+	default:
+		// Error if wrong type.
+		return builtins.NewOperandTypeErr(1, x, "string", "object")
 	}
 
-	// Validate the query against the schema, erroring if there's an issue.
-	schema, err := loadSchema(string(rawSchema))
+	// Transform the ASTs into Objects.
+	queryASTValue, err := ast.InterfaceToValue(queryDoc)
 	if err != nil {
 		return err
 	}
-	if err := validateQuery(schema, queryDoc); err != nil {
-		return err
-	}
-
-	// Generate AST/errors for the GraphQL schema, since the query
-	// passed validation.
-	schemaDoc, err := parseSchema(string(rawSchema))
+	schemaASTValue, err := ast.InterfaceToValue(schemaDoc)
```

Comment:
After much testing, I realized that query/schema validation in `gqlparser` *mutates* the underlying GraphQL AST object (decorating it with a lot of clutter fields, mostly), and that by moving the logic for preparing our output Rego objects *above* the mutating function calls, we can avoid all that mess, and get outputs that match `parse_query` and `parse_schema`'s results exactly.

Reply:
Hmm that's a good catch.